### PR TITLE
latte-dock: unstable-2022-09-06 -> unstable-2023-03-31

### LINKED ISSUE #212828: Latte-dock clicking on items does nothing
### Describe the bug
When I click on any program in latte-dock, nothing happens. If I run `latte-dock -d` in a terminal, I get these warnings:
```
[Warning : 1:31:04.614614] - "file:///nix/store/041wmy35j2h9rchm50lqrqfrbl28v792-latte-dock-unstable-2022-09-06/share/plasma/plasmoids/org.kde.latte.plasmoid/contents/ui/task/TaskMouseArea.qml:194: TypeError: Property 'canPresentWindows' of object Backend(0x213e3f0) is not a function"
[Warning : 1:38:00.355355] - "file:///nix/store/041wmy35j2h9rchm50lqrqfrbl28v792-latte-dock-unstable-2022-09-06/share/plasma/plasmoids/org.kde.latte.plasmoid/contents/ui/task/TaskMouseArea.qml:194: TypeError: Type error"
[Warning : 1:39:04.055055] - "file:///nix/store/041wmy35j2h9rchm50lqrqfrbl28v792-latte-dock-unstable-2022-09-06/share/plasma/plasmoids/org.kde.latte.plasmoid/contents/ui/task/SubWindows.qml:55: ReferenceError: LegacyWinIdList is not defined"
```

### Notify maintainers
@ysndr @greydot 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.89, NixOS, 22.11 (Raccoon), 22.11.1777.cdead16a444`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by Zahrun**
###### Description of changes

Updated to latest git commit + fixed https://github.com/NixOS/nixpkgs/issues/212828
This is my first PR. Any comment is welcome.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/latte-dock/default.nix [hunk 1]
```diff
@@ -1,30 +1,32 @@
...
 
   patches = [
     ./0001-Disable-autostart.patch
+    ./0002-Detect-kde-version.patch
   ];
 
   postInstall = ''
     mkdir -p $out/etc/xdg/autostart
     cp $out/share/applications/org.kde.latte-dock.desktop $out/etc/xdg/autostart
+    wrapProgram $out/bin/latte-dock --prefix PATH : ${lib.makeBinPath [ kde-cli-tools ]}
```
Comment (robryk) [inline pkgs/applications/misc/latte-dock/default.nix:29]: Do I understand correctly that this is newly required? (i.e. that this worked beforehand and something changed in latte-dock that caused it to start using kde-cli-tools in the meantime)

File: pkgs/applications/misc/latte-dock/default.nix [hunk 2]
```diff
@@ -1,30 +1,32 @@
...
     owner = "plasma";
     repo = "latte-dock";
-    rev = "cd36798a61a37652eb549d7dfcdf06d2028eddc4";
-    sha256 = "sha256-X2PzI2XJje4DpPh7gTtYnMIwerR1IDY53HImvEtFmF4=";
+    rev = "159496edc0609c953a813a41844bbb4f0ee22db4";
+    sha256 = "sha256-WqAlnwD/Q7+OUf2MuJfsFkruYQnlYahRCm1p+1zu8eA=";
   };
 
-  buildInputs = [ plasma-framework plasma-wayland-protocols qtwayland xorg.libpthreadstubs xorg.libXdmcp xorg.libSM wayland ];
+  buildInputs = [ plasma-framework plasma-wayland-protocols qtwayland xorg.libpthreadstubs xorg.libXdmcp xorg.libSM wayland kde-cli-tools ];
```
Comment (robryk) [inline pkgs/applications/misc/latte-dock/default.nix:16]: Do you need kde-cli-tools in buildInputs? If the build process doesn't use them (and the only way they're used is the `lib.makeBinPath` below) then this is not needed.

**Review by robryk — COMMENTED** [review]
Please format the commit message like [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) requests: `latte-dock: unstable-2022-09-06 -> unstable-2023-01-26`.

If you can find one, point at a changelog that describes what changed in the package in that interval in the commit message.

File: pkgs/applications/misc/latte-dock/default.nix (hunk 2 above)
Comment (Zahrun) [reply pkgs/applications/misc/latte-dock/default.nix:16]: It is not used in the build process, only at execution time. I will remove it from buildInputs.

File: pkgs/applications/misc/latte-dock/default.nix (hunk 1 above)
Comment (Zahrun) [reply pkgs/applications/misc/latte-dock/default.nix:29]: If you install the latte-dock version currently available in the Nix repo, you will see that clicking on an app icon in the dock will not do anything, which is kind of a core feature that it should open an app or show the window when clicking on an app. I described the issue here: https://github.com/NixOS/nixpkgs/issues/212828. The problem is that latte-dock is trying to detect the plasmashell version by parsing the output of `plasmashell -v` at run time. I tried to wrap plasma-workspace with latte-dock but to not avail, I think due to plasmashell segfaulting on nixos:
```
[aroun@nixos:~]$ plasmashell -v
plasmashell 5.26.4
Segmentation fault (core dumped)
```
So I wrapped kde-cli-tools instead, and added the /0002-Detect-kde-version.patch to replace `plasmashell -v` by `kbroadcastnotification -v`, which does not segfault and results in fixing the issue of apps not opening. I also thought that maybe wrapping kde-cli-tools would be lighter than wrapping plasma-workspace.
```
[:~] % kbroadcastnotification -v
kbroadcastnotification 5.26.4
[:~] %
```

**Comment by Zahrun** [comment]
> Please format the commit message like [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) requests: `latte-dock: unstable-2022-09-06 -> unstable-2023-01-26`.
> 
> If you can find one, point at a changelog that describes what changed in the package in that interval in the commit message.

Thank you for your review. I made the changes accordingly.

File: pkgs/applications/misc/latte-dock/default.nix (hunk 1 above)
Comment (robryk) [reply pkgs/applications/misc/latte-dock/default.nix:29]: Why was it checking `plasmashell` in the first place? Do you think someone just picked a random KDE thing, or is there a reason why they picked `plasmashell` and not e.g. `kbroadcastnotification`?

File: pkgs/applications/misc/latte-dock/default.nix (hunk 2 above)
Comment (robryk) [reply pkgs/applications/misc/latte-dock/default.nix]: JFYI, the point of `buildInputs` is that they are "present" when the build is running. They become runtime dependencies (i.e. part of the closure of the output) iff the output of the build contains references to paths within them.

**Review by robryk — APPROVED** [review]

**Review by robryk — CHANGES_REQUESTED** [review]
Actually, please wait a bit.

I'm more confused about this version. Version _of what_ do we need here? Of the environment this is running in? If so, then this is the wrong thing to do: that environment might be completely different from the version that we could depend on now (think system vs user profiles in NixOS or non-NixOS distributions).

**Comment by Zahrun** [comment]
Well latte-dock in it’s code has checks about what version of plasmashell it is running on. The line which was failing was this https://invent.kde.org/plasma/latte-dock/-/blob/master/plasmoid/package/contents/ui/task/TaskMouseArea.qml#L197 where it applies for plasma version < 22. Other code is used for plasma version > 23 and yet another for > 25. You can see here that they have many variables for plasma versions https://invent.kde.org/plasma/latte-dock/-/blob/master/plasmoid/package/contents/ui/main.qml#L49.

I don’t know how it works when it’s not running on plasma shell. I think it was designed around plasma in the first place. I know it can run on other shells as well but I don’t know how they do that in the code. I thought it was safe to use kbroadcastnotification since it returns the same version as plasmashell. But I can be wrong. Also using plasmashell -v did fail, I think due to the segfault.

**Comment by robryk** [comment]
Either way, this will at best be an imperfect approximation of the version of plasmashell that's actually running.

Is `kbroadcastnotification` usually present in the environment from which latte-dock is started? This sounds like one of the few situations where calling _that_ one is actually what we want. If so, the only reason to disprefer the original behaviour (calling `plasmashell`) is the crash.

So, about the crash: Is the `plasmashell` crash reproducible (it doesn't happen to me in `nix-shell -p plasmashell` using nixpkgs as of 40ef3bb63d4b6cf66aeb8468a0e8349cbc5aa5ca) for you? There was a recent update to plasma (b27befbb8c4b0afa7cf44c57cf748ef39d112d23); maybe that fixed the crash?

**Comment by Zahrun** [comment]
It seems the segfault is only present in NixOS 22.11
```
⌁65% [:~] 51s % nix-shell -p plasma-workspace -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/nixos-22.11.tar.gz --run "plasmashell -v"   
plasmashell 5.26.4
/tmp/nix-shell-59106-0/rc: line 3: 62394 Segmentation fault      (core dumped) plasmashell -v
⌁64% [:~] 5s 139 % nix-shell -p plasma-workspace -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/nixos-unstable.tar.gz --run "plasmashell -v"
plasmashell 5.26.5
⌁64% [:~] 4s % nix-shell -p plasma-workspace -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/master.tar.gz --run "plasmashell -v"      
plasmashell 5.26.5
⌁64% [:~] 5s %
```
So I changed latte-dock to include `plasma-workspace` instead of `kde-cli-tools`. Now I cannot really test it unless I update my whole system to unstable or master, since I am running NixOS stable, the different QT versions can't be run at the same time.
`[Fatal : 11:32:34.3535] - "Cannot mix incompatible Qt library (5.15.7) with this library (5.15.8)"`.

**Comment by robryk** [comment]
> So I changed latte-dock to include plasma-workspace instead of kde-cli-tools.

This still doesn't fix the issue caused by using a _different version_ of `plasma-workspace` than the one actually running. Consider a non-NixOS user who is running plasma-workspace provided by their distribution. In that case we'd want to look at the version of _that_ plasma-workspace, as opposed to the version that's available in nixpkgs.

**Comment by Zahrun** [comment]
> > So I changed latte-dock to include plasma-workspace instead of kde-cli-tools.
> 
> This still doesn't fix the issue caused by using a _different version_ of `plasma-workspace` than the one actually running. Consider a non-NixOS user who is running plasma-workspace provided by their distribution. In that case we'd want to look at the version of _that_ plasma-workspace, as opposed to the version that's available in nixpkgs.

I was able to confirm that the replacement of kde-cli-tools by plasma-workspace and no patch work fine when plasmashell does not segfault, since the recent update and it work for me, in the use case that we are running kde plasma in NixOS from the same repo as latte-dock.
If Nix is used as a package manager in a different distribution, how can we check the plasmashell version? One option could be to parse the value of X-KDE-PluginInfo-Version in `/usr/share/xsessions/plasma.desktop` but then in NixOS this file is somewhere in the nix store. I am new to NixOS so I am not sure about what is the best approach here.

**Comment by Zahrun** [comment]
This is an other way to get the running version of plasmashell. Maybe this one would work in all situations?
```
⏚ [:~] % qdbus org.kde.plasmashell /MainApplication org.qtproject.Qt.QCoreApplication.applicationVersion
5.26.4
```

**Comment by robryk** [comment]
Do I get it right that the problem with the vanilla one is solely the crash in NixOS 22.11?

**Comment by robryk** [comment]
Hmmm.... I can't reproduce the crash:

```
$ nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/nixos-22.11.tar.gz -p plasma-workspace --run 'plasmashell -v'
plasmashell 5.26.4
```

Can you reproduce this crash on another machine?

**Comment by Zahrun** [comment]
> Do I get it right that the problem with the vanilla one is solely the crash in NixOS 22.11?

No, even in unstable when plasmashell does not crash, https://github.com/NixOS/nixpkgs/issues/212828 persists, unless using the version in this branch/PR

**Comment by Zahrun** [comment]
Since the issue comes from https://invent.kde.org/plasma/latte-dock/-/blob/master/declarativeimports/core/environment.cpp#L65, I think the issue is only that latte-dock does not have access to the path to plasmashell. How do we give the path to another package?

**Comment by Zahrun** [comment]
https://github.com/NixOS/nixpkgs/issues/212828 has been properly fixed upstream.

So this PR is now just an update to the latest git commit with the added build dependency 'plasma-workspace'.

I would backport this to 22.11, but I'll leave that up to you.

**Pull request merged**